Zend/tests/type_declarations/union_types/class_loading.phpt

```diff
@@ -0,0 +1,62 @@
+--TEST--
+Class loading issues related to multiple class types in a union
+--FILE--
+<?php
+
+class Test {
+    public X|Y|Z|int $prop;
+}
+
+spl_autoload_register(function($class) {
+    echo "Loading $class\n";
+});
+
+$test = new Test;
+
+// Should not cause class loading, as not an object
+$test->prop = 42; 
+var_dump($test->prop);
+
+// Should try to load both classes
+try {
+    $test->prop = new stdClass;
+} catch (TypeError $e) {
+    echo $e->getMessage(), "\n";
+}
+
+if (true) {
+    class X {}
+}
+
+// Should not cause class loading, as X is already loaded
+try {
+    $test->prop = new X;
+} catch (TypeError $e) {
+    echo $e->getMessage(), "\n";
+}
+var_dump($test->prop);
+
+if (true) {
+    class Z {}
+}
+
+// TODO: Should this load class Y or not?
```

Comment:
Probably only if all previously loaded types don't match (optimisation)

Reply:
I'm inclined to agree that it should work that way. I think we need to acknowledge that it is undefined whether the load happens or not though and this would just be an optimization. E.g. in some other cases (like `Foo|callable`) I wouldn't want to guarantee that `Foo` is not loaded if `callable` can be proven without it, as it will complicate the implementation for little practical benefit.

Reply:
Note also that autoloading as a primary development tool will slowly fade with the advent of preloading, so undefined behavior isn't that big of an issue yet. Full autoloading (and skipping this optimisation for now) would be acceptable, having that kept in mind.

Reply:
Turns out the question is kinda moot, because this shouldn't be triggering autoloading in the first place. Fixed in https://github.com/php/php-src/commit/f8413887816c4cecda75e6ae59dc3d1a74bb4f77. May become an issue in the future if we do decide to make it load.